controller/src/main/java/cern/jarrace/controller/rest/controller/AgentContainerController.java

```diff
@@ -93,6 +96,33 @@ public String runService(@PathVariable(CONTAINER_NAME_VARIABLE_NAME) String cont
         throw new IllegalArgumentException("Service name must exist");
     }
 
+    @RequestMapping(value = "/{" + CONTAINER_NAME_VARIABLE_NAME + "}/read", method = RequestMethod.GET)
+    public ResponseEntity<String> readSource(@PathVariable(CONTAINER_NAME_VARIABLE_NAME) String containerName,
+                                             @RequestParam("class") String className) {
+        return agentContainerManager.findAgentContainer(containerName)
+                .map(container -> {
+                    try {
+                        return JarReader.ofContainer(container, reader -> {
+                            final String entry = className + JAVA_CLASS_SUFFIX;
+                            try {
+                                return ResponseEntity.ok(reader.readEntry(entry));
+                            } catch (NoSuchElementException e) {
+                                return ResponseEntity.badRequest()
+                                        .body("No class source found for entry " + entry);
+                            } catch (IOException e) {
+                                return ResponseEntity.status(INTERNAL_SERVER_ERROR)
+                                        .body("Failed to read entry " + entry + " inside container " + containerName);
+                            }
+                        });
+                    } catch (IOException e) {
+                        LOGGER.warn("Failed to read from container file {}: " + containerName, e);
+                        return ResponseEntity.status(INTERNAL_SERVER_ERROR)
+                                .body("Failed to open container of name " + containerName);
+                    }
+                })
```

Comment:
also you can put the different responses creation in their own, meaningfully named methods, it will improve understandability